File: shared/src/androidMain/kotlin/com/commuto/interfacemobile/kmService/kmTypes/PublicKey.kt
package com.commuto.interfacemobile.kmService.kmTypes

import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DERNull
import org.bouncycastle.asn1.pkcs.PKCSObjectIdentifiers
import org.bouncycastle.asn1.x509.AlgorithmIdentifier
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.MGF1ParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.security.PublicKey as JavaSecPublicKey
import javax.crypto.Cipher
import javax.crypto.spec.OAEPParameterSpec
import javax.crypto.spec.PSource

/**
 * The PublicKey class is a wrapper around the java.security.PublicKey class, with support for Commuto Interface IDs and
 * several methods for encoding the wrapped public key. The wrapped public key shall be that corresponding to a 2048-bit
 * RSA private key, and interfaceId is the SHA-256 hash of the public key encoded in PKCS#1 byte format.
 *
 * @property publicKey the java.security.PublicKey object around which this class is wrapped.
 * @property interfaceId the interface id derived from the public key
 */
class PublicKey {

    /**
     * Creates a PublicKey object using the PKCS#1 byte format of an RSA public key
     *
     * @param publicKeyBytes: the PKCS#1 byte encoded representation of the public key to be restored
     */
    constructor(publicKeyBytes: ByteArray) {
        val algorithmIdentifier = AlgorithmIdentifier(PKCSObjectIdentifiers.rsaEncryption,
            DERNull.INSTANCE)
        val pubKeyX509Bytes = SubjectPublicKeyInfo(algorithmIdentifier, publicKeyBytes).encoded
        val publicKey: JavaSecPublicKey = KeyFactory.getInstance("RSA")
            .generatePublic(X509EncodedKeySpec(pubKeyX509Bytes))
        this.publicKey = publicKey
        this.interfaceId = MessageDigest.getInstance("SHA-256")
            .digest(toPkcs1Bytes())
    }

    /**
     * Creates a PublicKey object using an RSA JavaSecPublicKey
     *
     * @param publicKey: the JavaSecPublicKey to be wrapped in a PublicKey
     */
    constructor(publicKey: JavaSecPublicKey) {
        this.publicKey = publicKey
        this.interfaceId = MessageDigest.getInstance("SHA-256")
            .digest(toPkcs1Bytes())
    }

    val interfaceId: ByteArray
    val publicKey: JavaSecPublicKey

    /**
     * Verifies a signature using this PublicKey's public key
     *
     * @param signedData: the data that was signed
     * @param signature: the signature
     *
     * @return Boolean: indicating the success of the verification
     */
    fun verifySignature(signedData: ByteArray, signature: ByteArray): Boolean {
        val signatureObj = Signature.getInstance("SHA256withRSA")
        signatureObj.initVerify(this.publicKey)
        signatureObj.update(signedData)
        return signatureObj.verify(signature)
    }

    /**
     * Encrypt the passed data using this PublicKey's RSA public key, using OEAP SHA-256 padding.
     *
     * @param clearData: the data to be encrypted
     * @return ByteArray: the encrypted data
     */
    fun encrypt(clearData: ByteArray): ByteArray {
        val encryptCipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding")
        val oaepParams = OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec("SHA-256"), PSource.PSpecified.DEFAULT)
        encryptCipher.init(Cipher.ENCRYPT_MODE, this.publicKey, oaepParams)
        return encryptCipher.doFinal(clearData)
    }

    /**
     * Encodes the RSA PublicKey to a PKCS1 formatted byte representation
     *
     * @return a ByteArray containing the PKCS1 representation of the PublicKey object
     */
    fun toPkcs1Bytes(): ByteArray {
        val pubKeyX509Bytes = this.publicKey.encoded
        val sPubKeyInfo: SubjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(pubKeyX509Bytes)
        val pubKeyPrimitive: ASN1Primitive = sPubKeyInfo.parsePublicKey()
        val pubKeyPkcs1Bytes: ByteArray = pubKeyPrimitive.encoded
        return pubKeyPkcs1Bytes
    }
}

File: iosApp/iosApp/kmService/kmTypes/PublicKey.swift
//
//  PublicKey.swift
//  iosApp
//
//  Created by jimmyt on 11/28/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import CryptoKit
import Foundation

//TODO: static method to re-create public key from raw bytes

/**
 * The PublicKey class is a wrapper around the SecKey class, with support for Commuto Interface IDs.
 * The wrapped public key shall be that corresponding to a 2048-bit RSA private key key, and interfaceId
 * is the SHA-256 hash of the public key encoded in PKCS#1 byte format.
 */
struct PublicKey {
    
    let interfaceId: Data
    let publicKey: SecKey
    
    /**
        Initializes a new PublicKey object, deriving interfaceId from the passed public key
     - Parameter publicKey: the public key to be wrapped
     */
    init(publicKey: SecKey) throws {
        self.publicKey = publicKey
        var error: Unmanaged<CFError>?
        guard let pubKeyBytes = SecKeyCopyExternalRepresentation(publicKey, &error) else {
            throw error!.takeRetainedValue() as Error
        }
        let interfaceIdDigest = SHA256.hash(data: pubKeyBytes as NSData)
        var interfaceIdByteArray = [UInt8]()
        for byte: UInt8 in interfaceIdDigest.makeIterator() {
            interfaceIdByteArray.append(byte)
        }
        let interfaceIdBytes = Data(bytes: interfaceIdByteArray, count: interfaceIdByteArray.count)
        self.interfaceId = interfaceIdBytes
    }
    
    /**
     * Verifies a signature using this KeyPair's public key
     *
     * - Parameter signedData: the data that was signed
     * - Parameter signature: the signature
     *
     * - Returns Boolean: indicating the success of the verification
     */
    func verifySignature(signedData: Data, signature: Data) throws -> Bool {
        let algorithm: SecKeyAlgorithm = .rsaSignatureMessagePKCS1v15SHA256
        var error: Unmanaged<CFError>?
        let success = SecKeyVerifySignature(publicKey, algorithm, signedData as CFData, signature as CFData, &error)
        if !success {
            throw error!.takeRetainedValue() as Error
        }
        return success
    }
    
    /**
     Encrypt the passed data using this PublicKey's RSA public key, using OEAP SHA-256 padding.
     - Parameter clearData: the data to be encrypted
     - Returns Data: the encrypted data
     */
    func encrypt(clearData: Data) throws -> Data {
        let algorithm: SecKeyAlgorithm = .rsaEncryptionOAEPSHA256
        var error: Unmanaged<CFError>?
        guard let cipherData = SecKeyCreateEncryptedData(publicKey, algorithm, clearData as CFData, &error) as Data? else {
            throw error!.takeRetainedValue() as Error
        }
        return cipherData
    }
    
}
